### ISSUE TITLE
Lightroom's "Reload Plugin" button does not work when MIDI2LR is opened

### ISSUE THREAD

**Issue by mathieucarbou** [rsjaffe/MIDI2LR#342]
LR freezes if I do not close mIDI2LR window before clicking on Reload Plugin.

I always had the issue, but this is quite annoying ;-) I was wondering if we can do something about it

**Comment by rsjaffe** [comment]
I think that's the attempt to close/restart the app. Does it still freeze if you disable closing the app when exiting lightroom?

If this continues to be a problem, we need to look at how ShutDown.lua and ShutDdownPlugin.lua are called during plugin reload and application shsutdown.

**Comment by mathieucarbou** [comment]
Actually, for my test, I had already uncheck the option to let MIDI2LR opened at shutdown. So yes the issue happens when the checkbox is unchecked, and LR freezes, until I manually close MIDI2LR: at this points, everything gets unlocked and the plugin restarts correctly.

**Comment by rsjaffe** [comment]
I've added a "stop server" option to the File ⇢ Plug-in Extras menu. Try that before reloading the plugin. If that works, we're going to have to relook at the plugin shutdown code to see if we can stop the server before the plugin stops. I suspect there is a Juce bug related to the pipe between LR and the app.

See https://github.com/rsjaffe/MIDI2LR/tree/feature/add_menu_item_shutdown_server for the test version with the shutdown option.

**Comment by mathieucarbou** [comment]
Hi @rsjaffe,
So yes, it works. It takes a little time but stop closed MIDI2LR and I was able to reload.

But is there a way to reload the plugin without the need to close MIDI2LR ?

Also is it nortmal that in `Shutdown.lua` for mac, there's a kill that is done instead of passing `--LRSHUTDOWN` ti the app ?

**Comment by rsjaffe** [comment]
The kill is in Shutdown.lua because the Mac app "hangs" and goes to 100% CPU utilization when LR closes. Sending the Mac app a message to close didn't help. The kill message is a temporary solution until we fix shutdown. I did a lot of things to fix shutdown (e.g., using smart pointers) and really debugged that stage of the app--at least in Windows, so the shutdown in windows is, as far as the app can control it, well-ordered. I don't know how well the Juce code shuts down.

That's where my suspicions about the pipe code comes from.

**Comment by rsjaffe** [comment]
Version 1.1.0 is where the shutdown code was fixed for Windows.

**Comment by rsjaffe** [comment]
try changing ShutDownPlugin.lua to the following:

```
if ProgramPreferences.StopServerOnExit then
  MIDI2LR.SERVER:close()
  MIDI2LR.RUNNING = false
end
```

**Comment by mathieucarbou** [comment]
having the option unchecked I suppose ?

![2016-12-14 08_53_58-set midi2lr options](https://cloud.githubusercontent.com/assets/61346/21184532/e2bd2fc4-c1da-11e6-8025-f1cf1bc98136.png)

**Comment by mathieucarbou** [comment]
It works! So it is ` MIDI2LR.SERVER:close()` that blocks and prevents the plugin to be reloaded. 
And the reload is wow!! Pretty fast!! Before, I had to wait about 10sec because I needed to close MIDI2LR and the program was blocking before and took time to relaod after also.

**Comment by mathieucarbou** [comment]
I tested a little more because what is weird is that if we prevent the socket from closing, then it remains open I guess ? And the next reload of the plugin will just create a new socket to MIDI2LR again, so it will be cumulative ?
I reloaded the plugin several times just to confirm and it runs slower and slower.

Does MIDI2LR has some callback somewhere that could prevent a client to close its connection properly ?

**Comment by rsjaffe** [comment]
My impression is that LR fails to completely unload the plugin. This is why we've had to exit LR when changing plugin files, rather than just reloading. So LR ends up with multiple instances of the plugin running. I believe the problem is in the `while MIDI2LR.RUNNING` loops in Client.lua. I've tried several ways of signaling termination, but without success. That probably means that the client doesn't close either, as that only occurs when Client.lua loops terminate.  I've started working on this problem at https://github.com/rsjaffe/MIDI2LR/tree/feature/shut_down_from_plugin but obviously needs more work. Fixing this might even fix the Mac shutdown issue.
